src/useCustomCompareMemo/useCustomCompareMemo.ts

```diff
@@ -0,0 +1,33 @@
+import { useMemo, useRef } from 'react';
+
+import type { DependencyList } from 'react';
+
+type DepsAreEqual<Deps extends DependencyList> = (savedDeps: Deps, deps: Deps) => boolean;
+
+/**
+ * Like `useMemo` but with a callback to defines that factory function can be invoked.
+ *
+ * @param factory useMemo factory function
+ * @param deps useMemo dependency list
+ * @param depsAreEqual defines that factory function can be invoked
+ * @returns useMemo result
+ */
+export const useCustomCompareMemo = <Factory extends () => unknown, Deps extends DependencyList>(
+  factory: Factory,
+  deps: Deps,
+  depsAreEqual: DepsAreEqual<Deps>
```

Comment:
Better name this argument as `comparator` and reuse `IDependenciesComparator` type from `useCustomCompareEffect`.

Dont mind `I` prefix - all prefixes will be removed in close future.